### ISSUE TITLE
Fast Shot trait fix by Haenlomal

### ISSUE THREAD

**Issue by FakelsHub** [phobos2077/sfall#335]
It seems that the old fix from Haenlomal is not correct, we need to remove it and apply the proper fix for the Fast Shot trait.

**F2 behavior**: Fast Shot is applied if its attack range is > 2 and prohibits all called attacks. (this logically)
**F2 haenlomal fix**: Fast Shot is applied if its attack range is > 2 and allows called attacks if range <=2. (this not logically)

**New behavior**: Fast Shot apply to the types: GUNS & MELEE & THROWING weapons.

Ugh, I'm kind of confused about this.
It seems that FastShotFix=2 works most logically prohibiting all called attacks and giving a bonus -1 AP for all.

**Comment by NovaRain** [comment]
Funny, I was testing the fix and tweaking the code last week (I guess you saw _someone's_ post on NMA). We need to discuss this one a bit more before stating "renovating" what has been working for 10 years.

Actually my opinion is kinda opposite. Here's how Fast Shot trait works in Fallout 1 & 2:
FO1:

|Weapon/Attack types  |-1 AP cost|No aimed shots|
|---------------------|:--------:|:------------:|
|Ranged weapons       |**YES**   |**YES**       |
|Melee/Unarmed weapons|**YES**   |**YES**       |
|Unarmed attacks      |**NO**    |**YES**       |

FO2 vanilla:

|Weapon/Attack types  |-1 AP cost|No aimed shots|
|---------------------|:--------:|:------------:|
|Ranged weapons       |**YES**   |**YES**       |
|Melee/Unarmed weapons|**NO**    |**YES**       |
|Unarmed attacks      |**NO**    |**YES**       |

FastShotFix=1:

|Weapon/Attack types  |-1 AP cost|No aimed shots|
|---------------------|:--------:|:------------:|
|Ranged weapons       |**YES**   |**YES**       |
|Melee/Unarmed weapons|**NO**    |**NO**        |
|Unarmed attacks      |**NO**    |**NO**        |

FastShotFix=2:

|Weapon/Attack types  |-1 AP cost|No aimed shots|
|---------------------|:--------:|:------------:|
|Ranged weapons       |**YES**   |**YES**       |
|Melee/Unarmed weapons|**YES**   |**YES**       |
|Unarmed attacks      |**YES**   |**YES**       |

_NOTE: "Unarmed attacks" means no weapon in the item slot, e.g. bare hands._

#### FO1 in-game description:
* You don't have time to aim for a targeted attack, because you attack faster than normal people. It costs you one less action point to use **a weapon**.

#### FO1 manual (page 3-15):
* You cannot perform targeted shots, but **all weapons** have AP -1 to use.

#### FO2 in-game description:
* You don't have time to aim for a targeted attack, because you attack faster than normal people. It costs you one less action point for **guns and thrown weapons**.

#### FO2 manual:
1. (page 27) You will _not even be able to switch to targeted shot mode_ if you are a Fast Shot.
2. (page 42) If you have selected the Fast Shot trait for your character, you cannot make targeted shots, and you will _never see the targeted shot icon_.
3. (page 130) Characters that have the Fast Shot trait will _never even see the targeted shot icon_.

If we go by the manual, the FastShotFix was never needed because the vanilla behavior did as what the manual said, and `=2` was never the real "Fallout 1 behavior" (it's largely due to my stupidity when writing descriptions for all those hidden settings in 3.6-3.7).

The main idea of the fix is about applying the effects _equally_: if you attack faster than normal people, you don't have time to aim (no aimed shots); but if you cannot attack faster, you should have time to aim like normal people do.
Later games (Van Buren and New Vegas) also use the same logic to limit the effects only on ranged weapons.

I think the reason of checking the weapon range instead of subtype is because naturally all GUNS/THROWING weapons have range > 2 (I mean, it's kinda silly if you have a knife with range 10 hexes or a gun can only shoot 1-2 hexes), also the same logic was used in the game scripts (for example, Hoss' script in NCR), so it's easier to just check the weapon range instead of going through the subtype process, and there are melee/range hybrids like spear and throwing knife.

I have a WIP code that changes `=2` to check if the player has a weapon in the hand and it works, and also tried to check the subtype instead but had some trouble on the result.

My proposal is:
Keep the `=1` behavior (effects only on ranged weapons), which has been the default for past 10 years.
New FastShotFix=2 ("fixed" FO1 behavior with the effects applied equally and only on _weapons_):
|Weapon/Attack types  |-1 AP cost|No aimed shots|
|---------------------|:--------:|:------------:|
|Ranged weapons       |**YES**   |**YES**       |
|Melee/Unarmed weapons|**YES**   |**YES**       |
|Unarmed attacks      |**NO**   |**NO**       |

And change the current `=2` to `=3` as the _alternative_ behavior in FO1 style.

**Comment by FakelsHub** [comment]
Clarification: FO2 vanilla:
Ranged weapons - what weapon is this all has a range attack > 2?
Melee/Unarmed weapons - Unarmed? if Melee or Unarmed has a range attack >=3, -1ap will be applied to it

**Comment by FakelsHub** [comment]
> If we go by the manual, the FastShotFix was never needed because the vanilla behavior did what the manual said

I agree that in F2 this works correctly without any fixes.

**Comment by NovaRain** [comment]
> Here I do not quite agree, this is Trait, it should always deprive something regardless of the conditions, in this case it takes all the aimed attacks, and gives a decent bonus in speed for all ranged attack.

Well, this is more like a subjective opinion / personally preference IMO. You can try asking this on the forums and see what people think, but that's the reason this fix was made back in the days so at least there are some people think the trait effects should be applied equally.

> I think can remove Haenlomal without pain and apply a new fix.

If you mean replacing the range check with subtype check, maybe.
I think the only thing that would be affected is the lightsaber in RP, because it was set up as a laser gun with range 1.

**Comment by FakelsHub** [comment]
"Applying Fast Shot Trait Fix. (Alternative version)"

I would change this logic to the original F1, in my opinion it is more correct, and it can be applied in f2ettu

**Comment by NovaRain** [comment]
> 
> > "Applying Fast Shot Trait Fix. (Alternative version)"
> 
> I would change this logic to the original F1, in my opinion it is more correct, and it can be applied in f2ettu

Et Tu has already been using the current `=2` (the _alternative_ / `=3` in my proposal above) with CALCAPCOST hook doing +1 AP cost for unarmed attacks (bare hands) to emulate the exact behavior in FO1.

My idea for all the settings is:
1 - the trait effects are applied on ranged weapons only.
2 - the trait effects are applied on ranged+melee+unarmed **weapons**.
3 - the trait effects are applied on all weapons & attacks (the current `=2`).

One of the reasons is it's easier to use CALCAPCOST hook to change the AP costs if someone like "unequal" effects.

**Comment by FakelsHub** [comment]
I would do this:
FastShotFix=1 (New)
Attack types | -1 AP cost | No aimed shots
-- | -- | --
Ranged weapons | YES | YES
Melee weapons | YES | YES
Unarmed attacks | NO | NO

FastShotFix=2  (Original F1)
Attack types | -1 AP cost | No aimed shots
-- | -- | --
Ranged weapons | YES | YES
Melee weapons | YES | YES
Unarmed attacks | NO | YES

FastShotFix=3 (Alternative version  like F2)
Attack types | -1 AP cost | No aimed shots
-- | -- | --
Guns/Thrown weapons | YES | YES
Melee weapons | NO | YES
Unarmed attacks | NO | YES

**Comment by NovaRain** [comment]
You surely like unequal effects but I don't. Looks like we might not have a common agreement in this case.

Let me rephrase my ideas:
Three settings all have different limits on the "no aimed shots" negative effect.
Because it's much easier to change AP cost for weapons/attacks with CALCAPCOST hook.

**Comment by FakelsHub** [comment]
> My idea for all the settings is:

Make tables not clear from the text)

**Comment by FakelsHub** [comment]
> You surely like unequal effects but I don't. Looks like we might not have a common agreement in this case.

This is not my like/dislike this is an original behavior and it is logical for Trait

**Comment by NovaRain** [comment]
> 
> 
> > My idea for all the settings is:
> 
> Make tables not clear from the text)

Fine.

FastShotFix=1 (Still the default, "Haenlomal's fix"):

|Weapon/Attack types  |-1 AP cost|No aimed shots|
|---------------------|:--------:|:------------:|
|Ranged weapons       |**YES**   |**YES**       |
|Melee/Unarmed weapons|**NO**    |**NO**        |
|Unarmed attacks      |**NO**    |**NO**        |

New FastShotFix=2 (tweaked FO1 behavior):

|Weapon/Attack types  |-1 AP cost|No aimed shots|
|---------------------|:--------:|:------------:|
|Ranged weapons       |**YES**   |**YES**       |
|Melee/Unarmed weapons|**YES**   |**YES**       |
|Unarmed attacks      |**NO**   |**NO**       |

New FastShotFix=3 (the old "=2", alternative behavior):

|Weapon/Attack types  |-1 AP cost|No aimed shots|
|---------------------|:--------:|:------------:|
|Ranged weapons       |**YES**   |**YES**       |
|Melee/Unarmed weapons|**YES**   |**YES**       |
|Unarmed attacks      |**YES**   |**YES**       |

> This is not my like/dislike this is an original behavior and it is logical for Trait

And who said it's only logical that trait effects must always deprive something regardless of the conditions? :)

> I just want to make it work, not as crooked as it is now - not dependent on the attack distance.

I asked in my previous comment: replacing the range check with weapon subtype check to see if a weapon is ranged (GUNS/THROWING)? If yes then I'm fine with that, as long as it doesn't break stuff (other than RP's lightsaber because it's set up unusually).

**Comment by FakelsHub** [comment]
> I think the only thing that would be affected is the lightsaber in RP, because it was set up as a laser gun with range 1.

:/ It melee weapon???

**Comment by NovaRain** [comment]
> 
> 
> > I think the only thing that would be affected is the lightsaber in RP, because it was set up as a laser gun with range 1.
> 
> :/ It melee weapon???

Well, it's "supposed" to be a melee weapon (obviously, unless someone never saw StarWar movie), but in its proto it's actually a laser gun (using "single" attack mode) that has only range of 1.

**Comment by FakelsHub** [comment]
> New FastShotFix=3 (the old "=2", alternative behavior):

This is then better left on 2

**Comment by NovaRain** [comment]
> 
> 
> > New FastShotFix=3 (the old "=2", alternative behavior):
> 
> This is then better left on 2

OK, we can put the new tweaked FO1 behavior as 3 since it's a new addition.

**Comment by NovaRain** [comment]
> 
> 
> > New FastShotFix=3 (the old "=2", alternative behavior):
> 
> This is then better left on 2
> 
> `Melee/Unarmed weapons` - I am confused by your Melee/Unarmed these are different things.

Oh, that means the player have a "close range" (range <= 2) weapon in his hands, my previous discussion was based on the "weapon range" check. Of course if we switch to `item_w_subtype`, MELEE and UNARMED types will be separated.

**Comment by FakelsHub** [comment]
> BUT it's been the default behavior for a decade and also the default in Steam/GOG release (because of the bundled sfall).

I don't care what's used in Steam/GOG (it's already out of date)

I left... Let's think.

**Comment by NovaRain** [comment]
> 
> 
> > BUT it's been the default behavior for a decade and also the default in Steam/GOG release (because of the bundled sfall).
> 
> I don't care what's used in Steam/GOG (it's already out of date)
> 
> I left... Let's think.

Just saying what new players would get when they install the game, or existing unofficial patches/RP/TCs that use older version of sfall. I wouldn't really expect all people know/want to update sfall right after installing the game.

Maybe you ask on NC and I ask on NMA to see other people's opinions?

**Comment by FakelsHub** [comment]
Ask for opinions on the forum.

**Comment by Lexx2k** [comment]
Had a long talk about this with NovaRain some days ago, and this is indeed a very tricky area. Generally we agreed that this "Fix" was never needed in the first place, and Fast Shot works correct in both, Fallout 1 and 2... it's just balanced differently.

My thoughts right now:
Why not remove this fix completely and move everything into script? We already have the option to override trait effects in perks.ini ... So technically we could just script these effects and everyone can do it the way they want. This means one kinda useless option less in the ddraw.ini and there is no discussion needed about how to change these effects.

**Comment by FakelsHub** [comment]
scripts have one bad feature - the order of execution.
engine fixes are always in the lower layer.

**Comment by NovaRain** [comment]
Maybe change the option to accept a list of settings similar to the red dotted line feature like Wipe said on NMA, so people can config the trait effects in the way they want.

**Comment by FakelsHub** [comment]
Okay, for now, let's leave it at that.
but add a correct fix by default for the f2 behavior.
`If (weapon == guns/thrownig AND range attack > 1) then -1ap`

I'll think of something else later.

**Comment by FakelsHub** [comment]
> Maybe change the option to accept a list of settings similar to the red dotted line feature like Wipe said on NMA, so people can config the trait effects in the way they want.

I definitely don't want to do this with sfall code, only if it's done via a script.
If you want to do it, you can do it.

**Comment by burner1024** [comment]
I think that Haenlomal's version makes most sense.
But I do agree that it's better to check actual attack modes rather than range. If a modder wants to add an extra-long spear, let him, but it should abide by melee requirements.

**Comment by NovaRain** [comment]
> 
> 
> Okay, for now, let's leave it at that.
> but add a correct fix by default for the f2 behavior.
> `If (weapon == guns/throwing AND range attack > 1) then -1ap`

What about my junk pocket pistol that can only shoot 1 hex? Just kidding.
I did a simple test to replace the range check in `item_w_mp_cost_` and it seems working:
```asm
static void __declspec(naked) item_w_mp_cost_hook() {
    __asm {
        call fo::funcoffs::item_hit_with_;  // get pointer to weapon
        call fo::funcoffs::item_w_subtype_; // eax - weapon, edx - hit_mode
        retn; // eax - weapon subtype
    }
}

HookCall(0x478C97, item_w_mp_cost_hook);
```

**Comment by FakelsHub** [comment]
> I think that Haenlomal's version makes most sense.

@burner1024 
this is for you, as the main fan of breaking the original mechanics)))

**Comment by FakelsHub** [comment]
@NovaRain
I'll make a commit with the code later.

I would also like to add the original F1 behavior as FastShotFix=3
Weapon/Attack types | -1 AP cost | No aimed shots
-- | -- | --
Ranged weapons | YES | YES
Melee/Unarmed weapons | YES | YES
Unarmed attacks | NO | YES

Other variants for behavior, let people add via the script.

**Comment by NovaRain** [comment]
> 
> 
> @NovaRain
> I'll make a commit with the code later.
> 
> I would also like to add the original F1 behavior as FastShotFix=3
> Other variants for behavior, let people add via the script.

But it can already be done with =2 and a hook script that simply increases AP cost when the player has no weapon as in Et Tu.
